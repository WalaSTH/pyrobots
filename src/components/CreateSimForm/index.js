// Form input and validation
import * as Yup from "yup";
import { Form, Formik } from "formik";

// Custom components for form input with formik and MUI
import TextField from "../FormsUI/TextField";
import SubmitFormButton from "../FormsUI/SubmitFormButton";

// MUI components
import { Card, Grid, Typography } from "@mui/material";

const initialFormState = {
  rounds: "",
  robot_ids: [1, 2],
};

const formValidation = Yup.object().shape({
  rounds: Yup.number()
    .integer()
    .typeError("Must be a valid number")
    .positive("Must be a positive number")
    .max(10000, "Max number of rounds is 10000")
    .required("Required"),
});

export default function CreateSimForm({ onSubmit }) {
  return (
    <Formik
      initialValues={{ ...initialFormState }}
      validarionSchema={formValidation}
      onSubmit={onSubmit}
    >
      <Card
        variant="outlined"
        sx={{
          marginTop: 3,
          padding: 3,
          borderRadius: 3,
        }}
      >
        <Form>
          <Grid
            container
            spacing={2}
            sx={{
              display: "flex",
              justifyContent: "center",
            }}
          >
            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                marginTop: -0.5,
                marginBottom: -0.5,
              }}
            >
              <Typography component="h1" variant="h5">
                Create Simulation
              </Typography>
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                marginTop: -0.5,
                marginBottom: -0.5,
              }}
            >
              <TextField
                name="rounds"
                label="Rounds"
                placeholder="10000"
                autoComplete="off"
              />
            </Grid>

            <Grid
              item
              xs={12}
              sx={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                marginTop: -0.5,
                marginBottom: -0.5,
              }}
            >
              <SubmitFormButton>Run simulation</SubmitFormButton>
            </Grid>
          </Grid>
        </Form>
      </Card>
    </Formik>
  );
}
