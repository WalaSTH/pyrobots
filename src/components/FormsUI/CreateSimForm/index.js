// Form input and validation
import * as Yup from "yup";
import { Form, Formik } from "formik";

// Custom components for form input with formik and MUI
import TextField from "../TextField";
import SubmitFormButton from "../SubmitFormButton";
import SelectRobot from "../SelectRobot";

// MUI components
import { Card, FormHelperText, Grid, Typography } from "@mui/material";
import { useState } from "react";

const initialFormState = {
  rounds: "",
  robots: ["", "", "", ""],
};

const MIN_ROBOTS_ERROR = "Select at least 2 robots";

const formValidation = Yup.object().shape({
  rounds: Yup.number()
    .integer()
    .typeError("Must be a valid number")
    .positive("Must be a positive number")
    .max(10000, "Max number of rounds is 10000")
    .required("Required"),
  robots: Yup.array().test("minRobots", MIN_ROBOTS_ERROR, (array) => {
    return (
      array.filter(function (e) {
        return e;
      }).length > 1
    );
  }),
});

export default function CreateSimForm({ onSubmit }) {
  const [isSelectError, setIsSelectError] = useState(false);

  function handleError() {
    setIsSelectError(true);
  }

  return (
    <Card
      variant="outlined"
      sx={{
        marginTop: 3,
        padding: 3,
        borderRadius: 3,
      }}
    >
      <Formik
        initialValues={{ ...initialFormState }}
        validationSchema={formValidation}
        onSubmit={onSubmit}
      >
        <Form>
          <Grid container spacing={2}>
            <Grid
              item
              xs={12}
              sx={{
                textAlign: "center",
              }}
            >
              <Typography mt={1} mb={1} component="h1" variant="h5">
                Create Simulation
              </Typography>
            </Grid>

            <Grid item xs={12} mb={-1}>
              <Typography
                sx={{
                  fontSize: { xs: "0.9rem", md: "1rem" },
                }}
                color="primary"
              >
                Number of rounds
              </Typography>
            </Grid>

            <Grid item xs={12}>
              <TextField
                name="rounds"
                label="Rounds"
                placeholder="10000"
                autoComplete="off"
                required={true}
              />
            </Grid>

            <Grid item xs={12} mb={-1}>
              <Typography
                sx={{
                  fontSize: { xs: "0.9rem", md: "1rem" },
                }}
                color="primary"
              >
                Select robots
              </Typography>
            </Grid>

            {isSelectError ? (
              <Grid item xs={12}>
                <FormHelperText error={isSelectError}>
                  {MIN_ROBOTS_ERROR}
                </FormHelperText>
              </Grid>
            ) : null}

            <Grid item xs={12} mb={-1}>
              <SelectRobot
                name="robots[0]"
                getRobotName="1"
                handleError={handleError}
              >
                Select robot 1
              </SelectRobot>
            </Grid>

            <Grid item xs={12} mb={-1}>
              <SelectRobot
                name="robots[1]"
                getRobotName="1"
                handleError={handleError}
              >
                Select robot 2
              </SelectRobot>
            </Grid>

            <Grid item xs={12} mb={-1}>
              <SelectRobot
                name="robots[2]"
                getRobotName="1"
                handleError={handleError}
              >
                Select robot 3
              </SelectRobot>
            </Grid>

            <Grid item xs={12} mb={-1}>
              <SelectRobot
                name="robots[3]"
                getRobotName="1"
                handleError={handleError}
              >
                Select robot 4
              </SelectRobot>
            </Grid>

            <Grid item xs={12}>
              <SubmitFormButton>Run simulation</SubmitFormButton>
            </Grid>
          </Grid>
        </Form>
      </Formik>
    </Card>
  );
}
