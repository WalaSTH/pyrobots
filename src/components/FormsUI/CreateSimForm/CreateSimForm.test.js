import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { screen, render, act } from "@testing-library/react";
import CreateSimForm from "./";
import renderer from "react-test-renderer";

console.error = () => {};

describe("CreateSimForm", () => {
  it("Render component", () => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => render(<CreateSimForm />));
    screen.getByText(/Create Simulation/i);
    screen.getByText(/Number of rounds/i);
    screen.getByLabelText(/Rounds/i);
    const selectRobotButtons = screen.getAllByRole("button", {
      name: /robotSelect/i,
    });
    expect(selectRobotButtons).toHaveLength(4);
    selectRobotButtons.forEach((button) => {
      expect(button).toHaveTextContent(/Select robot/i, { exact: false });
    });
    screen.getByText(/Run simulation/i);
  });

  it("Matches Snapshot", () => {
    const component = renderer.create(<CreateSimForm />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("handleError functionality", () => {
    CreateSimForm.handleError = jest.fn();
    CreateSimForm.handleError();
    expect(CreateSimForm.handleError).toHaveBeenCalledTimes(1);
  });
});
