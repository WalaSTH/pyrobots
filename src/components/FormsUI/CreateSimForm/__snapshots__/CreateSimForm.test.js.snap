// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSimForm Matches Snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-xheubk-MuiPaper-root-MuiCard-root"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1fpckdt-MuiGrid-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-fdwiyl-MuiTypography-root"
        >
          Create Simulation
        </p>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1x5ow3h-MuiGrid-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-1018lnw-MuiTypography-root"
        >
          Number of rounds
        </p>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r0:"
            id=":r0:-label"
          >
            Rounds
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="off"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id=":r0:"
              name="rounds"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="10000"
              required={true}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Rounds
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1x5ow3h-MuiGrid-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-1018lnw-MuiTypography-root"
        >
          Select robots
        </p>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1x5ow3h-MuiGrid-root"
      >
        <div>
          <div
            className="MuiBox-root css-k008qs"
          >
            <button
              aria-label="robotSelect"
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth css-xsqe0f-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              Select robot 1
            </button>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1x5ow3h-MuiGrid-root"
      >
        <div>
          <div
            className="MuiBox-root css-k008qs"
          >
            <button
              aria-label="robotSelect"
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth css-xsqe0f-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              Select robot 2
            </button>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1x5ow3h-MuiGrid-root"
      >
        <div>
          <div
            className="MuiBox-root css-k008qs"
          >
            <button
              aria-label="robotSelect"
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth css-xsqe0f-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              Select robot 3
            </button>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1x5ow3h-MuiGrid-root"
      >
        <div>
          <div
            className="MuiBox-root css-k008qs"
          >
            <button
              aria-label="robotSelect"
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth css-xsqe0f-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              Select robot 4
            </button>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-189e9pj-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          Run simulation
        </button>
      </div>
    </div>
  </form>
</div>
`;
